File: Baekjoon/Class3/44.10026 적록색약.js
const inputFile = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const input = require("fs").readFileSync(inputFile, "utf8").trim().split("\n");

MOVE_DIR = [[-1, 0], [1, 0], [0, -1], [0, 1]]

class CountColorArea {
    #areaCount = 0;
    constructor(colors, isEqual) {
        const n = colors.length;
        const visited = Array.from({ length : n }, () => new Array(n).fill(false));
        for(let r = 0; r < n; r++)
            for(let c = 0; c < n; c++)
                if(!visited[r][c]) { // 처음 발견한 색
                    const current_color = colors[r][c];
                    this.#areaCount++; // 개수 세기
                    let bfsHead = 0; // 너비 우선 탐색
                    const bfs = [[r, c]];
                    visited[r][c] = true;
                    while(bfsHead < bfs.length) {
                        const [startR, startC] = bfs[bfsHead++];
                        for(const [moveR, moveC] of MOVE_DIR) {
                            const endR = startR + moveR;
                            const endC = startC + moveC;
                            if (endR < 0 || endR >= n || endC < 0 || endC >= n)
                                continue;
                            // 주변 같은 색 발견
                            if (!visited[endR][endC] && isEqual(current_color, colors[endR][endC])) {
                                bfs.push([endR, endC]);
                                visited[endR][endC] = true;
                            }
                        }
                    }
                }
    }
    get areaCount() {
        return this.#areaCount;
    }
}

const n = Number(input[0]);

const colors = [] // 격자 그래프
for(let i = 1; i <= n; i++)
    colors.push(input[i]);

const out = [
    new CountColorArea(colors, (color1, color2) => { // 일반 비교 함수
        return color1 === color2;
    }).areaCount,
    new CountColorArea(colors, (color1, color2) => { // 색맹 비교 함수
        return color1 === color2
        || color1 === "R" && color2 === "G"
        || color1 === "G" && color2 === "R";
    }).areaCount
]
console.log(out.join(" "));

File: Baekjoon/Class3/44.10026 적록색약.py
from collections import deque
import sys
input = sys.stdin.readline

MOVE_DIR = ((-1, 0), (1, 0), (0, -1), (0, 1))

class CountColorArea:
    def __init__(self, colors, is_equal):
        self._area_count = 0
        n = len(colors)
        visited = [[False] * n for _ in range(n)]
        for r in range(n):
            for c in range(n):
                if not visited[r][c]: #처음 발견한 색
                    current_color = colors[r][c]
                    self._area_count += 1 #개수 세기
                    bfs = deque([(r, c)]) #너비 우선 탐색
                    visited[r][c] = True
                    while bfs:
                        start_r, start_c = bfs.popleft()
                        for move_r, move_c in MOVE_DIR:
                            end_r = start_r + move_r
                            end_c = start_c + move_c
                            if end_r < 0 or end_r >= n or end_c < 0 or end_c >= n:
                                continue
                            #주변 같은 색 발견
                            if not visited[end_r][end_c] and is_equal(current_color, colors[end_r][end_c]):
                                bfs.append((end_r, end_c))
                                visited[end_r][end_c] = True
    @property
    def area_count(self):
        return self._area_count

n = int(input())
colors = [input().rstrip() for _ in range(n)] #격자 그래프
print(CountColorArea(colors,
    lambda color1, color2: #일반 비교 함수
        color1 == color2)
    .area_count, end=' ')
print(CountColorArea(colors,
    lambda color1, color2: #색맹 비교 함수
        color1 == color2 or color1 in "RG" and color2 in "RG")
    .area_count)